README.md
**bot.sh.ori, xpebot.cfg.ori 파일은 필히 bot.sh, xpebot.cfg 로 '복사' 하셔서 사용하세요 안그러면 소스 업데이트가 안됩니다.**

------

### 자주 묻는 질문

- 사용자들이 자주 접하게 되는 문제들은 아래 링크로 대체 하였습니다.

- **[QNA.md](https://github.com/acidpop/xpebot/blob/master/QNA.md)** 이 링크를 참고하세요

------

### 이번 업데이트 내역

**0.6.3 (2017-11-29)**

wol 명령 수행시 등록된 Device 가 없는 경우 등록 절차가 진행 되지 않는 오류 수정

File 수신시 파일 이름 Encoding 실패에 대한 예외 처리 추가

/torrentsearch 기능 삭제 (DSM 에서 봇을 강제 종료 시키는 문제가 있음)

Torrent Download 알림 메시지에 작은 따옴표가 있는 경우 Exception 나는 현상 수정

xpebot.cfg 의 NOTY_CHAT_ID 에 ,로 구분하여 알림 받을 대상을 여러명 지정 할 수 있도록 수정

------

### 자세한 변경 사항은 다음 링크 참조

- 변경 내역이 길어져서 changelog 파일을 별도로 분리 하였습니다.

- **[CHANGELOG.md](https://github.com/acidpop/xpebot/blob/master/CHANGELOG.md)** 이 링크를 참고하세요

------


#XPEnology 전용 Telegram BOT 프로젝트


##설치 방법

1. DSM 에 접속하여 패키지 센터에서 "Git Server", "Python Module", "Download Station" 을 설치 한다.

2. Download Station 을 실행 하여 대상 경로를 설정 한다.

3. DSM 의 제어판에서 터미널 및 SNMP 항목을 클릭 후 "SSH 서비스 활성화"를 체크한다.

   제어판에서 사용자 항목으로 이동 후 고급 탭에서 "사용자 홈 서비스 활성화"를 체크한다.

   적용 버튼을 누른다. 

4. SSH 접속 툴(putty 또는 xshell등)을 이용하여 XPEnology IP 주소로 접속한다.

    (5.2-5644 기준으로 ID 는 'root' 암호는 관리자 계정(admin)의 암호이다)

5. DSM 6.0 이상 사용자는 다음 링크를 참고하여 DB에 코드를 업데이트해야 한다.

    **[DSM 6.0 xpebot 설정하기](http://blog.acidpop.kr/240)**

6. SSH로 접속 하였다면 자신의 홈 디렉토리로 이동한다. 

    cd /var/services/homes/{자신의계정}
    
7. xpebot 을 설치 한다.

    1) git clone "https://github.com/acidpop/xpebot"

    2) wget "https://bootstrap.pypa.io/get-pip.py"
   
    3) python ./get-pip.py
   
    4) pip install telepot BeautifulSoup psycopg2 Pillow psutil
   
    5) cd xpebot
    
    6) cp bot.sh.ori bot.sh
    
    7) cp xpebot.cfg.ori xpebot.cfg

    **bot.sh.ori 파일과 xpebot.cfg.ori 파일은 필히 bot.sh, xpebot.cfg 로 복사 하여 사용**

8. bot.sh 파일을 수정한다.

    1) 3번째 줄의 DSM_ACCOUNT="admin" 이 부분에서 admin 을 자신의 계정으로 변경한다.
  
9. xpebot.cfg 환경 설정 파일을 변경한다.

    1) NOTY_CHAT_ID 는 Downlaod Staion 에서 다운로드 진행 현황 알림을 받을 사용자의 Chat ID 값을 입력한다. 여러명일 경우 ,(콤마)를 이용해 구분한다.
  
    2) DSM_ID 는 자신의 DSM ID 를 입력한다.
  
    3) BOT_TOKEN 은 Telegram 의 Bot Father 에서 /newbot 을 요청하여 BOT 생성 후 자신의 BOT TOKEN 값을 입력한다.

	4) VALID_USER 는 인증된 사용자의 chat_id 값을 입력한다. 여러명일 경우 ,(콤마)를 이용해 구분한다.

	5) NAVER_API 섹션에 발급 받은 CLIENT_ID_KEY 값과 CLIENT_SECRET_KEY 값을 입력한다.

	6) RSS_NEWS 섹션에 보고 싶은 뉴스의 RSS 주소를 입력한다.
    
    7) DATA 섹션에 data.go.kr 에서 발급 받은 서비스 키를 입력한다.

 

10. xpebot 을 다음 명령어로 실행 한다. (DSM 6.x 사용자의 경우 root 권한에서 실행)

    ```
      시작
      ./bot.sh start
     
      종료 
      ./bot.sh stop
      
      실행 여부 확인
      ./bot.sh chk
    ```

  

------

## BotFather 에 Command 등록 하기

1. /secommands 메시지 보내기

2. 설정 할 BOT 계정 선택

3. 다음 메시지 보내기

    ```
    tfreeca - Tfreeca 조회
    torkim - TorrentKim 조회
    airkorea - 통합대기 지수 조회
    torrentsearch - 토렌트 검색
    wol - Wake On Lan 기능
    weather - 동네 날씨 또는 전국 날씨 요약 정보 조회
    systeminfo - NAS System 리소스 조회
    en2ko - 영어 문장을 한글로 기계 번역
    ko2en - 한글 문장을 영어로 기계 번역
    shorturl - url 을 짧게 줄여주는 기능
    txt2voice - 한글 문장을 음성으로 변환
    news - RSS 뉴스 URL 조회
    namuwiki - Namu Wiki 조회
    gettfreeca - Tfreeca 조회 후 토렌트 파일 다운로드
    gettorrent - TorrentKim 조회 후 토렌트 파일 다운로드
    help - 도움말
    cancel - 모드 취소
    ```

4. Bot Father 에게 Success! Command list updated. /help 메시지가 오면성공


------
  
# 사용 방법

**/tfreeca**
- tfreeca 토렌트 검색

**/gettfreeca**
- tfreeca 토렌트 조회 후 Torrent 파일 받기

~~**/torrentsearch**~~
- ~~토렌트 검색~~

**/torkim**
- Torrent Kim 조회

**/gettorrent**
- Torrent Kim 조회 후 Torrent 파일 받기

**/weather**
- 동네 날씨 또는 전국 날씨 요약 정보 조회

**/wol**
- Wake On Lan 기능

**/addwol**
- Wake On Lan Device 등록

**/delwol**
- Wake On Lan Device 삭제

**/systeminfo**
- NAS System 리소스 조회

**/en2ko**
- 영어 문장을 한글로 기계 번역 (Naver Developers 에 등록된 키가 있어야 함)

**/ko2en**
- 한글 문장을 영어로 기계 번역 (Naver Developers 에 등록된 키가 있어야 함)

**/shorturl**
- url 을 짧게 줄여주는 기능 (Naver Developers 에 등록된 키가 있어야 함)

**/txt2voice**
- 한글 문장을 음성으로 변환

**/airkorea**
- 통합대기 지수 조회 (data.go.kr 에서 대기오염정보 조회 API 서비스 키가 필요 함)

**/namuwiki**
- Namu Wiki 조회

**/news**
- RSS 뉴스 URL 조회

**/cancel**
- 모드 취소

**/help**
- 도움말


------


CHANGELOG.md
### 변경 내역

**0.6.3 (2017-11-29)**

wol 명령 수행시 등록된 Device 가 없는 경우 등록 절차가 진행 되지 않는 오류 수정

File 수신시 파일 이름 Encoding 실패에 대한 예외 처리 추가

/torrentsearch 기능 삭제 (DSM 에서 봇을 강제 종료 시키는 문제가 있음)

Torrent Download 알림 메시지에 작은 따옴표가 있는 경우 Exception 나는 현상 수정

xpebot.cfg 의 NOTY_CHAT_ID 에 ,로 구분하여 알림 받을 대상을 여러명 지정 할 수 있도록 수정

------

**0.6.2 (2017-07-24)**

Torrent Kim 에서 Torrent 파일 다운로드 실패 오류 수정

------

**0.6.1 (2017-04-07)**

/torkim 을 사용시 잘못된 Torrent 파일로 오류 나오는 문제 수정

Torrent Kim Redirect URL 대응

------

**0.6.0 (2017-03-14)**

/tfreeca , /gettfreeca 지원

torrent kim 자료들의 성인광고가 많이 포함 되어 tfreeca 지원

tfreeca 의 검색엔진이 좋지 않아서 키워드를 정확하게 입력 하셔야 합니다

ex) 다큐 170314 --> 검색 실패
    다큐        --> 검색 성공

위와 같이 키워드 2개의 조합 검색이 안됩니다

------

**0.5.12 (2017-02-20)**

/torkim 명령 사용후 토렌트 선택시 일부 특수문자가 포함된 torrent 파일 다운로드가 실패하는 문제 수정

------

**0.5.11 (2017-02-15)**

/torkim 명령 사용시 간혹 Torrent 파일 이름을 제대로 가져오지 못하는 문제 수정

------

**0.5.10 (2017-02-10)**

/torkim 명령 사용시 검색 결과가 없을때 오류 나는 문제 수정

/torkim 명령 결과에 대한 전체 텍스트를 미리 전송 하고 Inline Keyboard 가 보이도록 변경

------

**0.5.9 (2017-01-25)**

/start 명령 수행시 지원하는 명령 키보드가 제대로 표시 되지 않던 문제 수정.

------

**0.5.8 (2017-01-22)**

torrentsearch 동작에 이상이 있는 현상을 수정 하였습니다.

torrent 관련 명령 사용시 봇으로 SIGTERM 이 오는데 해당 시그널을 무시 하도록 변경 하였습니다.

------

**0.5.7 (2017-01-20)**

torrentsearch 명령 수행 후 cancel 을 하지 않고 다른 명령 수행시 오류 발생 부분을 수정하였습니다.

airkorea 수행 시 밤 12시 인경우 python 오류를 수정하였습니다.

------

**0.5.6 (2017-01-02)**

DSM 6.x 전용 bot.sh 파일 추가

DSM 6.x 에서는 최초 설치 후 bot6.sh.ori 파일을 bot.sh 로 복사 하셔서 사용하시면 됩니다.

telepot 이 업데이트 되면서 ReplyKeyboardHide 이름이 ReplyKeyboardRemove 이름으로 변경 되어 소스상에 업데이트 하였습니다.

pip install telepot --upgrade 를 수행 하신 다음 xpebot 을 업데이트 하셔서 사용하시면 됩니다.

토렌트 파일 전송시 파일 이름에 특정 내용이 있을 경우 문자열 인코딩 실패 오류를 수정하였습니다.

------

**0.5.5 (2016-10-26)**

DownloadStation 에서 Torrent 제목에 대괄호([])가 포함 되어 있고 대괄호 사이 공백이 있는 경우 메시지 전송을 못하는 문제 수정

잘못된 Torrent 파일 업로드 시 오류 메시지 전송 하도록 수정 (봇 재시작 후 "/xpebotupdate" 메시지를 보내야 적용 됩니다)

마찬가지로 DSM 6.0 이상 사용자는 "http://blog.acidpop.kr/240" 블로그를 참조 하여 2.Create OR Replace function 코드를 적용 하셔야 합니다.

------

**0.5.4 (2016-10-18)**

Download Station 을 체크 하는 DB Function 업데이트가 있습니다.

※ xpebot 을 0.5.4 버전으로 처음 시작 하시는 분은 아래 과정이 필요 없습니다.

토렌트 파일이나 마그넷 링크 등록시 "파일 다운로드가 시작되었습니다." 라는 메시지 없이 다운로드가 완료 되었습니다 메시지만 오는 

경우가 있어 조건식을 업데이트 하였습니다.

git pull 로 소스 업데이트 하신 다음 봇을 재시작 하신 다음

텔레그램 봇에게 "/xpebotupdate" 메시지를 보내주시면 업데이트가 완료 됩니다.


단, DSM 6.0 이상 사용자는 "http://blog.acidpop.kr/240" 해당 링크를 참조 하여 DownloadStation 계정으로 로그인 한 다음 psql 에 로그인 하여 

"2. Create OR Replace function" 에 해당 하는 코드를 붙여 넣기 하셔서 실행 하시면 됩니다.

------

**0.5.3 (2016-10-17)**

TorrentKim Url 변경 적용 (/torkim 명령 이용시 Torrent 파일이 잘못 다운로드 되는 문제 수정)

------

**0.5.2 (2016-10-12)**

Magnet Link 검출 기능 추가.

Magnet 링크를 봇에게 보내면 Download Station 에 등록하는 기능

------

**0.5.1 (2016-08-29)**

DSM Version Check Bug 수정

------

**0.5.0 (2016-08-29)**

DSM 6.0 이상 버전에서 api 오류들 수정

6.0 이상에서는 다음 과정이 필요하다. (<a href="http://blog.acidpop.kr/240" target="_blank">http://blog.acidpop.kr/240</a>)

------

**0.4.7 (2016-07-20)**

그룹 채팅방에서는 커스텀 키보드가 보이지 않도록 변경

Torrent 파일을 봇에게 전달 하면 오류가 발생하던 문제 수정

------

**0.4.6 (2016-07-01)**

/torkim 또는 /gettorrent 명령 사용시 검색 결과가 없으면 exception 이 발생 하는 문제 수정

검색 결과가 없으면 "검색 결과가 없다는 결과 메시지 전송"

------

**0.4.5 (2016-06-27)**

/torkim 명령 사용시 유니코드 파일명이 정상적으로 다운로드 되지 않는 문제 수정

/airkorea 기능에서 미세먼지와 초미세먼지의 좋음, 나쁨 등급 기준치 수정

Bot 에서 .torrent 파일을 전송 하면 정상적으로 작동 하지 않던 문제 수정

------

**0.4.4 (2016-06-01)**

봇 명령시 나오는 사용자 키보드를 그룹 채팅방에서는 앞에 '/' 기호를 붙이도록 변경

Naver API 가 작동 하지 않던 문제 수정

tgbot.db 파일은 최종으로 업데이트 (이후부터는 수정 될 예정 없음)

------

**0.4.3 (2016-05-27)**

Group 대화방에서 Bot 작동 할 수 있도록 추가 개발

Group 대화방에 초대 후 /torrentsearch@xpebot

형태로 커맨드를 보낸 후 /var/log/xpebot.log 를 확인 하면 Group 대화방의 chat id 확인 가능

해당 Group ID 를 xpebot.cfg 의 invalid user 에 추가

그룹 대화방에서 봇은 "/" Slash 가 앞에 있어야만 메시지를 받을 수 있습니다

```
사용자	: /torrentsearch@xpebot
BOT		: 검색 할 Torrent 제목을 입력하세요
사용자	: /다큐멘터리
BOT		: 검색 결과 
```

위와 같은 방법으로 사용 가능하다.

------

**0.4.2 (2016-05-25)**

토렌트 파일 제목에 _ (underbar) 또는 * 문자가 들어가면 메시지를 보내지 못하는 문제 수정
tgbot.db 파일 git pull 할때 오류 시 

git checkout -- tgbot.db

위 명령으로 강제 업데이트 하실 수 있습니다.

tgbot.db 파일은 이 버전 이후에 되도록 업데이트가 없을 예정입니다.

------

**0.4.1 (2016-05-23)**
- tgbot.db 와 torrent 검색 결과를 저장하는 DB를 분리
- Torrent 검색 기능 로그 상세화
- /cancel 모드 도움말에 추가

------

**0.4 (2016-05-19)**
- 다음 패키지를 설치 해야 합니다. pip install Pillow
- Telepot 8.0 (BOT API 2.0) 에 맞춰 업데이트 하였습니다.  easy_install --upgrade telepot 명령을 이용하여 기존 telepot 사용자는 업데이트 하셔야 합니다. telepot 6.x 이하는 사용 불가
- torrent 파일 전송시 Watch Directory 에 파일을 다운로드 합니다. cfg 파일에서 watch_dir 키에 watch_directory 경로를 입력하세요.
- /airkorea - 미세 먼지 조회기능 추가 (data.go.kr 에서 '대기오염정보 조회' API에 대해 활용 신청 후 서비스 키가 필요합니다.)
- /torkim 추가 - TorrentKim 에서 Magnet 이 아닌 Torrent 파일을 검색하여 Watch Directory에 다운로드 하는 기능, 추천 수를 표시합니다 (TorrentKim 사정에 의해 작동이 안될 수 있습니다)
- /gettorrent 추가 - TorrentKim 에서 Magnet 이 아닌 Torrent 파일을 검색하여 사용자에게 파일을 전송 하는 기능 (TorrentKim 사정에 의해 작동이 안될 수 있습니다)
- /namuwiki - 나무 위키에서 키워드를 검색하여 해당 문서 링크를 찾아 주는 기능

/torkim, /gettorrent 명령은 이번에 추가된 BOT API 2.0 을 사용합니다. Inline Keyboard 를 이용하여 사용자에게 선택 버튼을 제공합니다.

xpebot.cfg.ori 와 bot.sh.ori 파일이 업데이트 되었습니다.

다시 복사 하셔서 사용하세요

내부적으로 사용하는 File DB가 업데이트 되었습니다.

tgbot.db 파일을 다른 이름으로 변경 후에 git 소스를 업데이트 하셔야 합니다.

기존에 등록된 WOL 데이터를 다시 등록 하셔야 합니다.

입력 시간 초과 기능은 삭제 되었습니다.

0.4 버전에서 Telepot 8.0 버전을 지원하며 새로운 기능들이 추가 되어서 작동이 안될 수 있습니다.

문의 사항이 있으시면 http://blog.acidpop.kr/notice/225 블로그 또는 github 의 issue 를 이용하여주세요.


------

**0.3 (2016-02-15)**
- bot.sh 에서 실행 경로 변경(/volume1/homes -> /var/service/homes)
- 날씨 정보 조회시 시,도,구가 다르고 동 이름이 같을 경우 요약 정보를 잘 못 보내던 문제 수정

bot.sh.ori 파일이 업데이트 되었습니다.

------

**0.2 (2016-02-13)**
- RSS 뉴스 리더 기능 추가
- NAS 의 시스템 리소스 조회 기능 추가(cpu, ram, disk)
- /start 명령 추가 (사용 가능 기능을 Custom Keyboard로 보여주기)

airkorea.py
#-*- coding: utf-8 -*-


import main
import BotManager
from LogManager import log

import os
import sys
import datetime
import time
from PIL import Image, ImageDraw, ImageFont
import requests
import linecache


# pip install Pillow

class airkorea(object):

    grade_color = { 0 : (0, 0, 0, 255),
                    1 : (78, 137, 246, 255),
                    2 : (91, 212, 100, 255),
                    3 : (254, 127, 65, 255),
                    4 : (249, 74, 75, 255)}

    sido_xy = { '서울': [104, 84], 
    			'인천': [40,107],
    			'강원': [194,84],
    			'경기': [122,116],
    			'충북': [164,133],
    			'충남': [82, 154],
    			'대전': [132, 177],
    			'경북': [222, 152],
    			'전북': [112, 210],
    			'대구': [185, 191],
    			'울산': [240, 198],
    			'광주': [62, 259],
    			'경남': [173, 245],
    			'부산': [236, 233],
    			'전남': [118, 278],
    			'제주': [75, 331]
    }
    
    sido_list = {'서울', '인천', '강원', '경기', '충북', '충남', '대전', '경북', '전북', '대구', '울산', '광주', '경남', '부산', '전남', '제주'}

    def PrintException(self):
        exc_type, exc_obj, tb = sys.exc_info()
        f = tb.tb_frame
        lineno = tb.tb_lineno
        filename = f.f_code.co_filename
        linecache.checkcache(filename)
        line = linecache.getline(filename, lineno, f.f_globals)
        log.error('EXCEPTION IN ({}, LINE {} "{}"): {}'.format(filename, lineno, line.strip(), exc_obj))
                
    # 통합대기 등급 0~50, 51~100, 101~250, 251~
    # 미세먼지 등급 0~30, 31~80, 81~150, 151~
    # 초미세먼지 등급 0~15, 16~50, 51~100, 101~
    def GetKHAIGradeColor(self, val):
        if val < 0:
            return self.grade_color[0]
        elif val >= 0 and val <= 50:
            return self.grade_color[1]
        elif val >= 51 and val <= 100:
            return self.grade_color[2]
        elif val >= 101 and val <= 250:
            return self.grade_color[3]
        else:
            return self.grade_color[4]

    def GetPM10GradeColor(self, val):
        if val < 0:
            return self.grade_color[0]
        elif val >= 0 and val <= 25:
            return self.grade_color[1]
        elif val >= 26 and val <= 50:
            return self.grade_color[2]
        elif val >= 51 and val <= 75:
            return self.grade_color[3]
        else:
            return self.grade_color[4]

    def GetPM25GradeColor(self, val):
        if val < 0:
            return self.grade_color[0]
        elif val >= 0 and val <= 15:
            return self.grade_color[1]
        elif val >= 16 and val <= 25:
            return self.grade_color[2]
        elif val >= 26 and val <= 50:
            return self.grade_color[3]
        else:
            return self.grade_color[4]


    def isNum(self, s):
        try:
            float(s)
            return True
        except ValueError:
            return False

    def ChangeTimeString(self, strTime):
        strTime = strTime.replace('24:', '00:')
        time1 = time.strptime(strTime, "%Y-%m-%d %H:%M")
        strDate = time.strftime("%Y.%m.%d %H:00 기준")
        return strDate


    def GetSidoAirData(self, sido):
        serviceKey = main.botConfig.GetDataServiceKey()
        url = "http://openapi.airkorea.or.kr/openapi/services/rest/ArpltnInforInqireSvc/getCtprvnRltmMesureDnsty?numOfRows=200&pageNo=1&sidoName=" + sido + "&ver=1.0&ServiceKey=" + serviceKey + "&_returnType=json"

        try:
            r = requests.get(url)

            if r.status_code != 200:
                return -1, -1, -1, ''

            response = r.json()

            items = response['list']

            khai_value = 0
            khai_count = 0

            pm10_value = 0
            pm10_count = 0

            pm25_value = 0
            pm25_count = 0

            dataTime = items[0]['dataTime']

            for item in items:
                if item.get('khaiValue') != None and self.isNum(item['khaiValue']):
                    khai = int(item['khaiValue'])
                    khai_value += khai
                    khai_count += 1

                if item.get('pm10Value') != None and self.isNum(item['pm10Value']):
                    pm10 = int(item['pm10Value'])
                    pm10_value += pm10
                    pm10_count += 1

                if item.get('pm25Value') != None and self.isNum(item['pm25Value']):
                    pm25 = int(item['pm25Value'])
                    pm25_value += pm25
                    pm25_count += 1

        except requests.exceptions.RequestException as e:
            log.info('Request Exception : %s', e)
            return -1, -1, -1, ''
        except:
            self.PrintException()
            return -1, -1, -1, ''

        return int(khai_value / khai_count), int(pm10_value / pm10_count), int(pm25_value / pm25_count), dataTime


    def SendAirKorea(self, bot, chat_id):
        execute_path = main.botConfig.GetExecutePath()

        font_path = execute_path + '/NanumGothicExtraBold.ttf'
        fnt = ImageFont.truetype(font_path, 14)

        font2_path = execute_path + '/NanumBarunGothic.ttf'
        fnt2 = ImageFont.truetype(font2_path, 12)

        img_map_path = execute_path + '/airmap.png'
        base = Image.open(img_map_path).convert('RGBA')
        khai_img = Image.new('RGBA', base.size, (255, 255, 255, 0))
        drow_khai = ImageDraw.Draw(khai_img)
    
        pm10_img = Image.new('RGBA', base.size, (255, 255, 255, 0))
        drow_pm10 = ImageDraw.Draw(pm10_img)
    
        pm25_img = Image.new('RGBA', base.size, (255, 255, 255, 0))
        drow_pm25 = ImageDraw.Draw(pm25_img)
    
        date = ''
    
        for sido in self.sido_list:
            khai, pm10, pm25, dataTime = self.GetSidoAirData(sido)
        
            if khai == -1 and pm10 == -1 and pm25 == -1:
                log.error('Get Sido Air Data Fail')
                return
        
            xy = self.sido_xy.get(sido)
            if xy == None:
                continue
            
            khai_val = str(khai)
            w, h = drow_khai.textsize(khai_val)
            x = xy[0] - (w / 2) - 1
            y = xy[1]
            drow_khai.text((x, y), khai_val, font=fnt, fill=self.GetKHAIGradeColor(int(khai_val)))
            
            pm10_val = str(pm10)
            w, h = drow_pm10.textsize(pm10_val)
            x = xy[0] - (w / 2) - 1
            y = xy[1]
            drow_pm10.text((x, y), pm10_val, font=fnt, fill=self.GetPM10GradeColor(int(pm10_val)))
            
            pm25_val = str(pm25)
            w, h = drow_pm25.textsize(pm25_val)
            x = xy[0] - (w / 2) - 1
            y = xy[1]
            drow_pm25.text((x, y), pm25_val, font=fnt, fill=self.GetPM25GradeColor(int(pm25_val)))

            date = self.ChangeTimeString(dataTime).decode('utf-8')

    

        drow_khai.text((15, 5), u'통합대기', font=fnt, fill=(0, 0, 0, 255))
        drow_khai.text((90, 5), date, font=fnt, fill=(0, 0, 0, 255))
        
        drow_pm10.text((15, 5), u'미세먼지', font=fnt, fill=(0, 0, 0, 255))
        drow_pm10.text((90, 5), date, font=fnt, fill=(0, 0, 0, 255))
    
        drow_pm25.text((15, 5), u'초미세먼지', font=fnt, fill=(0, 0, 0, 255))
        drow_pm25.text((90, 5), date, font=fnt, fill=(0, 0, 0, 255))
    
        # 통합대기 등급 0~50, 51~100, 101~250, 251~
        drow_khai.ellipse(((210,300),(223,312)), fill=(78, 137, 246, 255), outline="white")
        drow_khai.ellipse(((210,318),(223,330)), fill=(91, 212, 100, 255), outline="white")
        drow_khai.ellipse(((210,336),(223,348)), fill=(254, 127, 65, 255), outline="white")
        drow_khai.ellipse(((210,354),(223,366)), fill=(249, 74, 75, 255), outline="white")
        drow_khai.rectangle(((205, 296), (312, 368)), fill=None, outline=(190, 190, 190, 255))
    
        drow_khai.text((227, 300), u'좋음 0~50', font=fnt2, fill=(0, 0, 0, 255))
        drow_khai.text((227, 318), u'보통 51~100', font=fnt2, fill=(0, 0, 0, 255))
        drow_khai.text((227, 336), u'나쁨 101~250', font=fnt2, fill=(0, 0, 0, 255))
        drow_khai.text((227, 354), u'매우 나쁨 251~', font=fnt2, fill=(0, 0, 0, 255))
        drow_khai.text((5, 354), u'※ 출처 : 한국환경공단', font=fnt2, fill=(0, 0, 0, 255))
    
        # 미세먼지 등급 0~30, 31~80, 81~150, 151~
        # 미세먼지 등급 WHO 기준 50 이상은 나쁨
        drow_pm10.ellipse(((210,300),(223,312)), fill=(78, 137, 246, 255), outline="white")
        drow_pm10.ellipse(((210,318),(223,330)), fill=(91, 212, 100, 255), outline="white")
        drow_pm10.ellipse(((210,336),(223,348)), fill=(254, 127, 65, 255), outline="white")
        drow_pm10.ellipse(((210,354),(223,366)), fill=(249, 74, 75, 255), outline="white")
        drow_pm10.rectangle(((205, 296), (312, 368)), fill=None, outline=(190, 190, 190, 255))
    
        drow_pm10.text((227, 300), u'좋음 0~25', font=fnt2, fill=(0, 0, 0, 255))
        drow_pm10.text((227, 318), u'보통 26~50', font=fnt2, fill=(0, 0, 0, 255))
        drow_pm10.text((227, 336), u'나쁨 51~75', font=fnt2, fill=(0, 0, 0, 255))
        drow_pm10.text((227, 354), u'매우 나쁨 75~', font=fnt2, fill=(0, 0, 0, 255))
        drow_pm10.text((5, 354), u'※ 출처 : 한국환경공단', font=fnt2, fill=(0, 0, 0, 255))
    
    
        # 초미세먼지 등급 0~15, 16~50, 51~100, 101~
        # 초미세먼지 등급 WHO 기준 25 이상은 나쁨
        drow_pm25.ellipse(((210,300),(223,312)), fill=(78, 137, 246, 255), outline="white")
        drow_pm25.ellipse(((210,318),(223,330)), fill=(91, 212, 100, 255), outline="white")
        drow_pm25.ellipse(((210,336),(223,348)), fill=(254, 127, 65, 255), outline="white")
        drow_pm25.ellipse(((210,354),(223,366)), fill=(249, 74, 75, 255), outline="white")
        drow_pm25.rectangle(((205, 296), (312, 368)), fill=None, outline=(190, 190, 190, 255))
    
        drow_pm25.text((227, 300), u'좋음 0~15', font=fnt2, fill=(0, 0, 0, 255))
        drow_pm25.text((227, 318), u'보통 16~25', font=fnt2, fill=(0, 0, 0, 255))
        drow_pm25.text((227, 336), u'나쁨 26~50', font=fnt2, fill=(0, 0, 0, 255))
        drow_pm25.text((227, 354), u'매우 나쁨 51~', font=fnt2, fill=(0, 0, 0, 255))
        drow_pm25.text((5, 354), u'※ 출처 : 한국환경공단', font=fnt2, fill=(0, 0, 0, 255))
    
    
        out_khai = Image.alpha_composite(base, khai_img)
        out_pm10 = Image.alpha_composite(base, pm10_img)
        out_pm25 = Image.alpha_composite(base, pm25_img)
    
        out_khai.save('/tmp/air_khai.png')
        out_pm10.save('/tmp/air_pm10.png')
        out_pm25.save('/tmp/air_pm25.png')
            
        #sender.sendMessage(date)
        #sender.sendPhoto(open('/tmp/air_khai.png'))
        #sender.sendPhoto(open('/tmp/air_pm10.png'))
        #sender.sendPhoto(open('/tmp/air_pm25.png'))

        #bot = BotManager.BOTManager().GetBot()
        bot.sendMessage(chat_id, date)
        bot.sendPhoto(chat_id, open('/tmp/air_khai.png'))
        bot.sendPhoto(chat_id, open('/tmp/air_pm10.png'))
        bot.sendPhoto(chat_id, open('/tmp/air_pm25.png'))



